### PR TITLE
Raise RuntimeError when trying to use in-place functions in autograd

### PR DESCRIPTION
**Pull request by lvdmaaten**
Summary:
At present, the autograd logic performs in-place functions on `CrypTensor`s for which `requires_grad == True` without keeping track of the in-place operation. This is dangerous and can lead to unexpected behaviors.

After looking into PyTorch's logic for dealing with in-place functions in autograd, I decided the effort needed to support this in CrypTen would be substantial and the advantages would be limited.

This diff makes sure a `RunTimeError` is raised whenever the user performs an in-place operation on a tensor that is tracked by the autograd.

Differential Revision: D24385646

### CODE REVIEW

File: crypten/gradients.py [hunk 1]
```diff
@@ -36,11 +36,14 @@ def register_function_cls(cls):
...
     Returns gradient function for the CrypTen function with the specified name.
+    Also returns a boolean indicating whether or not the function is in-place.
     """
     if name in FUNCTION_REGISTRY:
-        return FUNCTION_REGISTRY[name]
-    else:
-        return None
+        return FUNCTION_REGISTRY[name], False
+    elif name.endswith("_"):
+        if name[:-1] in FUNCTION_REGISTRY:  # this is an in-place function
```
Comment (gmuraru) [inline crypten/gradients.py:44]: NITPICK: Using ```name[:-1]``` will create a new object.
Maybe it is worth saving it in a variable and using that variable in the ```if statement``` and the ```return```?

Also, something cool that I did not know about Python.
It seems there is something called [memoryview](https://docs.python.org/3/library/stdtypes.html#memoryview) that lets you look at the data without creating a copy of the element:
```
mem_view_name = memoryview(name.encode())
print(str(bytes(mem_view_name[:-1])))
```

PS: I think using ```memoryview``` might be a little bit excessive to use in this scenario :\

**Review by gmuraru — APPROVED** [review]
